File: thrift/perf/py/test_server.py
#!/usr/local/bin/python2.6 -tt
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements. See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership. The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License. You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied. See the License for the
# specific language governing permissions and limitations
# under the License.
#
from __future__ import absolute_import

import optparse
import sys

from apache.thrift.test.load import LoadTest
from apache.thrift.test.sync_load_handler import LoadHandler
from thrift.protocol.TBinaryProtocol import TBinaryProtocolAcceleratedFactory
from thrift.protocol.THeaderProtocol import THeaderProtocol
from thrift.protocol.THeaderProtocol import THeaderProtocolFactory
from thrift.transport.THeaderTransport import THeaderTransport
from thrift.transport import TSocket, TSSLSocket
from thrift.transport import TTransport
from thrift.server import TServer, TNonblockingServer, \
    TProcessPoolServer, THttpServer

def main():
    op = optparse.OptionParser(usage='%prog [options]', add_help_option=False)
    op.add_option('-p', '--port',
                  action='store', type='int', dest='port', default=1234,
                  help='The server port')
    op.add_option('-s', '--servertype',
                  action='store', type='string', dest='servertype',
                  default='TGeventServer',
                  help='Type name of server')
    op.add_option('-w', '--num_workers',
                  action='store', type='int', dest='workers', default=4,
                  help='Number of worker processes')
    op.add_option('-Q', '--max_queue_size',
                  action='store', type='int', dest='max_queue_size', default=0,
                  help='Max queue size, passed to TNonblockingServer')
    op.add_option('-h', '--header',
                  action='store_true', help='Use the generated ContextIface')
    op.add_option('-?', '--help',
                  action='help',
                  help='Show this help message and exit')

    (options, args) = op.parse_args()
    if args:
        op.error('trailing arguments: ' + ' '.join(args))

    handler = LoadHandler()
    if options.servertype == 'TGeventServer':
        # only import TGeventServer when necessary. TGeventServer calls
        # monkey_patch, which breaks other servers
        from apache.thrift.test.sync_load_handler import GeventLoadHandler
        from thrift.server import TGeventServer
        handler = GeventLoadHandler()
    processor = LoadTest.Processor(handler)

    if options.header:
        pfactory = THeaderProtocolFactory(True, \
                                          [THeaderTransport.HEADERS_CLIENT_TYPE,
                                           THeaderTransport.FRAMED_DEPRECATED,
                                           THeaderTransport.UNFRAMED_DEPRECATED,
                                           THeaderTransport.HTTP_CLIENT_TYPE])
        if options.servertype == 'TNonblockingServer':
            print 'Header transport, backwards compatible with framed'
        else:
            print 'Header transport, backwards compatible with ' + \
                'unframed, framed, http'
    else:
        if options.servertype == 'TNonblockingServer':
            print 'Framed transport'
        else:
            print 'Unframed transport'
        pfactory = TBinaryProtocolAcceleratedFactory()

    transport = TSocket.TServerSocket(options.port)
    tfactory = TTransport.TBufferedTransportFactory()
    if options.servertype == "TNonblockingServer":
        server = TNonblockingServer.TNonblockingServer(processor, transport,
                            pfactory, maxQueueSize=options.max_queue_size)
    elif options.servertype == "TProcessPoolServer":
        server = TProcessPoolServer.TProcessPoolServer(processor, transport,
                                                       tfactory,
                                                       pfactory)
        print 'Worker processes: ' + str(options.workers)
        server.setNumWorkers(options.workers)
    elif options.servertype == "TGeventServer":
        print 'Worker processes: ' + str(options.workers)
        # Gevent makes its own server transport.
        server = TGeventServer.TGeventServer(options.port,
                                             processor, None,
                                             tfactory,
                                             pfactory)
        server.setNumWorkers(options.workers)
    else:
        ServerClass = getattr(TServer, options.servertype)
        server = ServerClass(processor, transport, tfactory, pfactory)

    print 'Serving ' + options.servertype + \
        ' requests on port %d...' % (options.port,)
    server.serve()

if __name__ == '__main__':
    rc = main()
    sys.exit(rc)

File: thrift/lib/hs/tests/TestServer.hs
--
-- Licensed to the Apache Software Foundation (ASF) under one
-- or more contributor license agreements. See the NOTICE file
-- distributed with this work for additional information
-- regarding copyright ownership. The ASF licenses this file
-- to you under the Apache License, Version 2.0 (the
-- "License"); you may not use this file except in compliance
-- with the License. You may obtain a copy of the License at
--
--   http://www.apache.org/licenses/LICENSE-2.0
--
-- Unless required by applicable law or agreed to in writing,
-- software distributed under the License is distributed on an
-- "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
-- KIND, either express or implied. See the License for the
-- specific language governing permissions and limitations
-- under the License.
--

{-# LANGUAGE OverloadedStrings #-}
module TestServer
       ( runServer
       ) where

import Control.Exception
import Network
import System.IO
import qualified Data.HashMap.Strict as Map
import qualified Data.Text.Lazy as Text

import ThriftTest
import ThriftTest_Iface
import ThriftTest_Types

import Thrift
import Thrift.Server

data TestHandler = TestHandler
instance ThriftTest_Iface TestHandler where
  testVoid _ = return ()
  testString _ = return
  testByte _ = return
  testI32 _ = return
  testI64 _ = return
  testDouble _ = return
  testStruct _ = return
  testNest _ = return
  testMap _ = return
  testStringMap _ = return
  testSet _ = return
  testList _ = return
  testEnum _ = return
  testTypedef _ = return
  testMapMap _ _ = return $ Map.fromList
    [ (-4, Map.fromList [ (-4, -4)
                        , (-3, -3)
                        , (-2, -2)
                        , (-1, -1)
                        ])
    , (4,  Map.fromList [ (1, 1)
                        , (2, 2)
                        , (3, 3)
                        , (4, 4)
                        ])
    ]
  testInsanity _ x = return $ Map.fromList
    [ (1, Map.fromList [ (TWO  , x)
                       , (THREE, x)
                       ])
    , (2, Map.fromList [ (SIX, default_Insanity)
                       ])
    ]
  testMulti _ byte i32 i64 _ _ _ = return Xtruct
    { xtruct_string_thing = Text.pack "Hello2"
    , xtruct_byte_thing   = byte
    , xtruct_i32_thing    = i32
    , xtruct_i64_thing    = i64
    }
  testException _ s = case s of
    "Xception"   -> throw $ Xception 1001 s
    "TException" -> throw ThriftException
    _ -> return ()

  testMultiException _ s1 s2 = case s1 of
    "Xception"   -> throw $ Xception 1001 "This is an Xception"
    "Xception2"  -> throw $ Xception2 2002 default_Xtruct
    "TException" -> throw ThriftException
    _ -> return default_Xtruct{ xtruct_string_thing = s2 }

  testOneway _ _ = return ()

runServer :: Protocol p => (Handle -> p Handle) -> Socket -> IO ()
runServer p = runThreadedServer (accepter p) TestHandler ThriftTest.process
  where
    accepter p s = do
      (h, _, _) <- accept s
      return (p h, p h)
